inc/saitunnel.h

```diff
@@ -0,0 +1,444 @@
+ *
+ *  \{
+ */
+
+/**
+ * @brief Enum defining tunnel map types.
+ */
+typedef enum _sai_tunnel_map_type_t
+{
+    /** TUNNEL Map overlay ECN to underlay ECN   */
+    SAI_TUNNEL_MAP_OECN_TO_UECN = 0x00000001,
+
+    /** TUNNEL Map underlay ECN to overlay ECN   */
+    SAI_TUNNEL_MAP_UECN_TO_OECN = 0x00000002,
+
+    /** TUNNEL Map VNI to VLAN ID  */
+    SAI_TUNNEL_MAP_VNI_TO_VLAN_ID = 0x00000003,
+
+    /** TUNNEL Map VLAN ID to VNI */
+    SAI_TUNNEL_MAP_VLAN_ID_TO_VNI = 0x00000004,
+
+    /* -- */
+    /* Custom range base value */
+    SAI_TUNNEL_MAP_CUSTOM_RANGE_BASE = 0x10000000
+
+} sai_tunnel_map_type_t;
+
+typedef enum _sai_tunnel_map_attr_t
+{
+    /** tunnel Map type [sai_tunnel_map_type_t] (MANDATORY_ON_CREATE|CREATE_ONLY) */
+    SAI_TUNNEL_MAP_ATTR_TYPE = 0x00000000,
+
+    /** tunnel mapper  [sai_tunnel_map_list_t] */
+    SAI_TUNNEL_MAP_ATTR_MAP_TO_VALUE_LIST = 0x00000001,
+
+    /* -- */
+    /* Custom range base value */
+    SAI_TUNNEL_MAP_ATTR_CUSTOM_RANGE_BASE = 0x10000000
+
+} sai_tunnel_map_attr_t;
+
+/**
+ * @brief Create tunnel Map
+ *
+ * @param[out] tunnel_map_id tunnel Map Id
+ * @param[in] attr_count number of attributes
+ * @param[in] attr_list array of attributes
+ *
+ * @return  SAI_STATUS_SUCCESS on success
+ *          Failure status code on error
+ */
+typedef sai_status_t(*sai_create_tunnel_map_fn)(
+    _Out_ sai_object_id_t* tunnel_map_id,
+    _In_ uint32_t attr_count,
+    _In_ const sai_attribute_t *attr_list
+    );
+
+/**
+ * @brief Remove tunnel Map
+ *
+ *  @param[in] tunnel_map_id tunnel Map id to be removed.
+ *
+ *  @return  SAI_STATUS_SUCCESS on success
+ *           Failure status code on error
+ */
+typedef sai_status_t(*sai_remove_tunnel_map_fn) (
+    _In_  sai_object_id_t   tunnel_map_id
+    );
+
+/**
+ * @brief Set attributes for tunnel map
+ *
+ * @param[in] tunnel_map_id tunnel Map Id
+ * @param[in] attr attribute to set
+ *
+ * @return  SAI_STATUS_SUCCESS on success
+ *          Failure status code on error
+ */
+typedef sai_status_t(*sai_set_tunnel_map_attribute_fn)(
+    _In_ sai_object_id_t  tunnel_map_id,
+    _In_ const sai_attribute_t *attr
+    );
+
+/**
+ * @brief  Get attrbutes of tunnel map
+ *
+ * @param[in] tunnel_map_id  tunnel map id
+ * @param[in] attr_count  number of attributes
+ * @param[inout] attr_list  array of attributes
+ *
+ * @return SAI_STATUS_SUCCESS on success
+ *        Failure status code on error
+ */
+typedef sai_status_t(*sai_get_tunnel_map_attribute_fn)(
+    _In_ sai_object_id_t   tunnel_map_id,
+    _In_ uint32_t attr_count,
+    _Inout_ sai_attribute_t *attr_list
+    );
+
+typedef enum _sai_tunnel_type_t
+{
+    SAI_TUNNEL_IPINIP,
+
+    SAI_TUNNEL_IPINIP_GRE,
+
+    SAI_TUNNEL_IPINIP_GRE_AND_KEY,
+
+    SAI_TUNNEL_VXLAN,
+
+    SAI_TUNNEL_MPLS,
+
+} sai_tunnel_type_t;
+
+typedef enum _sai_tunnel_ttl_mode_t
+{
+    SAI_TUNNEL_TTL_COPY_FROM_INNER,
+
+    SAI_TUNNEL_TTL_USER_DEFINE
+
+} sai_tunnel_ttl_mode_t;
```

Comment:
the sai_tunnel_ttl_mode_t is current define for the encap, we probably need separate it into sai_tunnel_encap_ttl_mode and sai_tunnel_decap_ttl_mode